### ISSUE TITLE
Refresh token not received for Android App

### ISSUE THREAD

**Issue by dobrinsky** [DuendeSoftware/Support#1158]
**Which version of Duende.AccessTokenManagement are you using?**

6.0.8

**Which version of .NET are you using?**

6.0

**Describe the bug**

We have problems with the application both on the web version and the Android one where the login expires after 60 minutes and we need to login again. As we have read online, we should receive a refresh token that we are not receiving. There are big chances that this is because of the configuration, but we cannot seem to find the issue.

**To Reproduce**

We have created a project with the default Authentication Angular project template in Visual Studio and use it in the browser.

Create an Android project.

**Expected behavior**

We should receive a refresh token which on the Android project would give the user a seamless experience with the authentication process.

**Log output/exception with stacktrace**

The configuration of our application is: 

.NET client configuration: 

```
"AndroidID": {
        "Profile": "IdentityServerSPA",
        "Enabled": true,
        "ClientId": "AndroidID",
        "RequireClientSecret": false,
        "ClientName": "AndoidApp",
        "ClientSecrets": [ { "Value": "" } ],
        "RequirePkce": true,
        "AllowedGrantTypes": [ "code", "implicit" ],
        "AllowedScopes": [ "openid", "profile", "email", "phone", "api",  "offline_access"],
        "RedirectUris": [ "https://website/well-known/openid-configuration" ],
        "RequireConsent": false,
        "AllowOfflineAccess": true,
        "AccessTokenLifetime": 7200,
        "IdentityTokenLifetime": 7200
      }
```

Android:

```
{
  "client_id": "AndroidID",
  "redirect_uri": "https://website/authentication/login-callback",
  "client_secret": "secret",
  "authorization_scope": "openid SIGADAPI profile",
  "discovery_uri": "https://website/.well-known/openid-configuration",
  "authorization_endpoint_uri": "",
  "token_endpoint_uri": "",
  "registration_endpoint_uri": "",
  "response_type": "code",
  "https_required": false,
  "end_session_redirect_uri": "https://website/authentication/logout-callback",
  "offline_access": true
}
```

**Comment by RolandGuijt** [comment]
A refresh token is only returned when the offline_access scope is requested.
It looks like you have enabled offline_access on both the identity server and the android side but the scope is not actually requested.

Can you try changing "authorization_scope" on the Android side like this:

`"authorization_scope": "openid SIGADAPI profile offline_access",`

Please let me know if that helps. If not we can investigate further.

BTW: What framework are you using on the Android side?

**Comment by dobrinsky** [comment]
The above configuration was the only one we managed to use in order to obtain the login token in the first place. 

If we try with what you suggested @RolandGuijt (as we did initially), we get an error of type:

https://website/home?errorId=CfDJ8CFq3xaSUyNFlKyzL25qxkzli9mqpoEgwCon29vdT9PsmZH_RVwbrvI3_c2yzIlWCl8B-LBQ9qzWT71jY74I0JtW_yMeF8LoiSbXhYG45ar0gK-dCysAGYcBByOq3dTZ1yYnD6XXtgLtCpYpnPk7R91Wmhpg6kcmCLXhHrH7LcnI0ESV2bZM8QVQfoLwBmOiojkgMV9_lJ8LM9AWLDznXQ0rVnQ7mSw-dt1kjVYB_vimQ-d0uRf61Q2GNF0VYaR9Z6CZDjQEDdJWMZqNWZ04h7gKjeszup6APQghRrYxL6POQlqtCM5bN_8aA8BMR5RFPd6XrDBhNe4xv5IsZXfq906kY8keCFGejiwa7eJPtQUeMQ7gG6TIF4CtoN8rzYKd5bXhFtwvxXlxUc8W6IW0xiywx6UIhtYjecw6kiWlA4ScTep1tEQugHewOPu1r0f5LeVOKR-xs-37Sjt4gWz6UKmhGyupZtLr6mdv3mfOI-gfigv2F7zNJJX63WDkAevcrojUkY_wAoUyd0swKluD5OKSZIX49xlPFl-uDsQYYe9ZPiDGqyAFjAiu8hnarht2WQ 

and we are not getting redirected to the login page.

As for the Android library, we are using https://github.com/openid/AppAuth-Android on Android side

**Comment by dobrinsky** [comment]
In Program.cs we have added the logger with more information for authentication as it is written in the documentation:

Log.Logger = new LoggerConfiguration()
    .MinimumLevel.Debug()
    .MinimumLevel.Override("Microsoft", LogEventLevel.Warning)
    .MinimumLevel.Override("Microsoft.Hosting.Lifetime", LogEventLevel.Information)
    .MinimumLevel.Override("System", LogEventLevel.Warning)
    .MinimumLevel.Override("Microsoft.AspNetCore.Authentication", LogEventLevel.Information)
    .WriteTo.File("errorlog.txt", rollingInterval: RollingInterval.Day)
    .CreateLogger();

but nothing is logged. If tehre are any C# exceptions, they are logged, but nothing in the Authentication process. 

We are open to anything: code share, meetings, any possible solutions.

Also, in the Visual Studio project template the access token has a very short expiry time and the user experience is drastically impacted by this. We have one component that should display the stock of wastes which is permanently displayed, but after a little time, it expires.

**Comment by RolandGuijt** [comment]
So as soon as you're requesting the offline_access scope you get the error?

Hard to see what's wrong from this side of the keyboard.

Can you try our [Maui sample client](https://github.com/DuendeSoftware/Samples/tree/main/various/clients/Maui/MauiApp1) against your identity provider and see what happens?
And/or try some other client (like and asp.net core one) and request the offline_access scope?

**Comment by dobrinsky** [comment]
yes, as soon as we add the offline_access, we get that error. 

We will try the example and the basic template fresh, maybe there is a difference or a bug that has been solved since

**Comment by dobrinsky** [comment]
But I think the problem is with some configuration in the Duende, something I was not able to find until now. In the browser also, the authentication is expires and the behavior is not the best. If I close the browser tab and then open it again with another link, I need to click the "Login" button, it does not see the user as logged in. 

Is there any possibility to set the expiration of the token to a longer time (e.g. one year)?

**Comment by axelalex13** [comment]
Hi @RolandGuijt , 

I work with @dobrinsky and tested the Maui sample with our identity provider. 

I updated the OidcClient with our configuration and the result is the same: if we use the Scope = "openid SIGADAPI profile", we can receive the access token, but if we add the offline_access scope, the browser opens but the login page is not displayed and the URL contains an erorrId just like @dobrinsky showed before.

**Comment by RolandGuijt** [comment]
The Visual Studio template isn't the best for your scenario.
Please use our [upgrade guide](https://docs.duendesoftware.com/identityserver/v7/upgrades/spa_to_duende/) to set Duende IdentityServer up in the right way.

**Comment by RolandGuijt** [comment]
@dobrinsky The upgrade guide is basically starting from scratch. If that's not possible, to get to the bottom of the problem logging has to work so we need to figure out why it's not working. 
The configuration looks OK. But you mentioned you don't see any messages about authentication in the file?

You also mentioned the lifetime of the token being very short. But in the client config I see that it is set to 7200 seconds, 2 hours. If the lifetime is less than that then maybe IdentityServer can find/read the client config. Where is the client configured? In appsettings.json?

**Comment by dobrinsky** [comment]
Yes, the client is configured in appsettings.json

I started from yesterday to follow the QuickStarts and probably, in time, if everything works, we will migrate the code (functionalities) to the newly project with the correct configuration, but this will take time.

I have a question: for an angular project, the JavaScript client application QuickStart is a good example to follow?

**Comment by RolandGuijt** [comment]
I think that is by far the better approach in the long term.
Do you need any further help with the current situation or are you now focusing on the new setup? If you still want to troubleshoot the current solution you could check the access token lifetime and if the client config is read correctly like I suggested earlier.

For an Angular project, the quickstart is great, but keep in mind that is using BFF. The browser doesn't see any tokens with BFF which is the recommended approach for SPAs these days. [This](https://docs.duendesoftware.com/identityserver/v7/bff/overview/) is a good place to start reading if you're not familiar.

**Comment by dobrinsky** [comment]
I already changed the lifetime of the token to test the functionalities. We will get back with the results.

Before Duende, I managed to create the authentication mechanism for the angular project before Microsoft created the template that we used to migrate the code because that initial implementation was not the best one, it also had hiccups and there was no refresh. The initial authentication mechanism was with JWT token as well, I think it should be manageable, if  there are any problems, I will get back if it is ok
